File: Sources/App/Utils/Markdown.swift
#if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
	import Darwin
#elseif os(Linux)
	import Glibc
#endif

import Foundation

/* Kuyawa - 2016/30/12. Used in regex for linux, also in StringUtils.swift

- Linux compatibility:
  Uses a Typealias for NSRegularExpression
  Uses an extension to TextCheckingResult

*/

#if os(Linux)
typealias NSRegularExpression = RegularExpression
typealias NSTextCheckingResult = TextCheckingResult
extension TextCheckingResult {
	func rangeAt(_ n: Int) -> NSRange {
		return self.range(at: n)
	}
}
#endif


extension String {
    /* Already defined in StringUtils
    func match(_ pattern: String) -> Bool {
        guard self.characters.count > 0 else { return false }
        if let first = self.range(of: pattern, options: .regularExpression) {
            let match = self.substring(with: first)
            return !match.isEmpty
        }
        
        return false
    }
    */

    /* Already defined in StringUtils
    func trim() -> String {
        return self.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    */
    
    func remove(_ pattern: String) -> String {
        guard self.characters.count > 0 else { return self }
        if let first = self.range(of: pattern, options: .regularExpression) {
            return self.replacingCharacters(in: first, with: "")
        }
        
        return self
    }
    
    func prepend(_ text: String) -> String {
    	if text.isEmpty { return self }
        return text + self
    }
    
    func append(_ text: String) -> String {
    	if text.isEmpty { return self }
        return self + text
    }
 
    func enclose(_ fence: (String, String)?) -> String {
        return (fence?.0 ?? " ") + self + (fence?.1 ?? " ")
    }
}

extension NSMutableString {
    func matchAndReplace(_ rex: String, _ rep: String, options: NSRegularExpression.Options? = []) {
        if let regex = try? NSRegularExpression(pattern: rex, options: options!) {
	        let range = NSRange(location: 0, length: self.length)
	        _ = regex.replaceMatches(in: self, options: [], range: range, withTemplate: rep)
	    } else {
	    	print("Regex not valid")
	    }
    }
}


class Markdown {
    
    func parse(_ text: String) throws -> String {
    	print("Markdown enter...")
        var md = NSMutableString(string: text)
        
        cleanHtml(&md)
        parseHeaders(&md)
        parseBold(&md)
        parseItalic(&md)
        parseDeleted(&md)
/*
        parseImages(&md)
        parseLinks(&md)
        parseUnorderedLists(&md)
        parseOrderedLists(&md)
        parseBlockquotes(&md)
        parseCodeBlock(&md)
        parseCodeInline(&md)
        parseHorizontalRule(&md)
        parseYoutubeVideos(&md)
        parseParagraphs(&md)
*/        
        return String(describing: md)
    }
    
    func cleanHtml(_ md: inout NSMutableString) {
        md.matchAndReplace("<.*?>", "")
    }
    
    func parseHeaders(_ md: inout NSMutableString) {
        md.matchAndReplace("^###(.*)?", "<h3>$1</h3>", options: [.anchorsMatchLines])
        md.matchAndReplace("^##(.*)?", "<h2>$1</h2>", options: [.anchorsMatchLines])
        md.matchAndReplace("^#(.*)?", "<h1>$1</h1>", options: [.anchorsMatchLines])
    }

    func parseBold(_ md: inout NSMutableString) {
        md.matchAndReplace("\\*\\*(.*?)\\*\\*", "<b>$1</b>")
    }
    
    func parseItalic(_ md: inout NSMutableString) {
        md.matchAndReplace("\\*(.*?)\\*", "<i>$1</i>")
    }
    
    func parseDeleted(_ md: inout NSMutableString) {
        md.matchAndReplace("~~(.*?)~~", "<s>$1</s>")
    }
    
/*    
    func parseImages(_ md: inout NSMutableString) {
        md.matchAndReplace("!\\[(\\d+)x(\\d+)\\]\\((.*?)\\)", "<img src=\"$3\" width=\"$1\" height=\"$2\" />")
        md.matchAndReplace("!\\[(.*?)\\]\\((.*?)\\)", "<img alt=\"$1\" src=\"$2\" />")
    }
    
    func parseLinks(_ md: inout NSMutableString) {
        md.matchAndReplace("\\[(.*?)\\]\\((.*?)\\)", "<a href=\"$2\">$1</a>")
        md.matchAndReplace("\\[http(.*?)\\]", "<a href=\"http$1\">http$1</a>")
        md.matchAndReplace("(^|\\s)http(.*?)(\\s|\\.\\s|\\.$|,|$)", "$1<a href=\"http$2\">http$2</a>$3 ", options: [.anchorsMatchLines])
    }
    
    func parseUnorderedLists(_ md: inout NSMutableString) {
        //md.matchAndReplace("^\\*(.*)?", "<li>$1</li>", options: [.anchorsMatchLines])
        parseBlock(&md, format: "^\\*", blockEnclose: ("<ul>", "</ul>"), lineEnclose: ("<li>", "</li>"))
    }
    
    func parseOrderedLists(_ md: inout NSMutableString) {
        parseBlock(&md, format: "^\\d+[\\.|-]", blockEnclose: ("<ol>", "</ol>"), lineEnclose: ("<li>", "</li>"))
    }
    
    func parseBlockquotes(_ md: inout NSMutableString) {
        //md.matchAndReplace("^>(.*)?", "<blockquote>$1</blockquote>", options: [.anchorsMatchLines])
        parseBlock(&md, format: "^>", blockEnclose: ("<blockquote>", "</blockquote>"))
        parseBlock(&md, format: "^:", blockEnclose: ("<blockquote>", "</blockquote>"))
    }
    
    func parseCodeBlock(_ md: inout NSMutableString) {
        md.matchAndReplace("```(.*?)```", "<pre>$1</pre>", options: [.dotMatchesLineSeparators])
        //parseBlock(&md, format: "^\\s{4}", blockEnclose: ("<pre>", "</pre>"))
    }
    
    func parseCodeInline(_ md: inout NSMutableString) {
        md.matchAndReplace("`(.*?)`", "<code>$1</code>")
    }
    
    func parseHorizontalRule(_ md: inout NSMutableString) {
        md.matchAndReplace("---", "<hr>")
    }
    
    func parseYoutubeVideos(_ md: inout NSMutableString) {
        md.matchAndReplace("\\[youtube (.*?)\\]", "<p><a href=\"http://www.youtube.com/watch?v=$1\" target=\"_blank\"><img src=\"http://img.youtube.com/vi/$1/0.jpg\" alt=\"Youtube video\" width=\"240\" height=\"180\" /></a></p>")
    }
    
    func parseParagraphs(_ md: inout NSMutableString) {
        md.matchAndReplace("\n\n([^\n]+)\n\n", "\n\n<p>$1</p>\n\n", options: [.dotMatchesLineSeparators])
    }
    
    func parseBlock(_ md: inout NSMutableString, format: String, blockEnclose: (String, String), lineEnclose: (String, String)? = nil) {
        let lines = md.components(separatedBy: .newlines)
        var result = [String]()
        var isBlock = false
        var isFirst = true
        
        for line in lines {
            var text = line
            if text.match(format) {
                isBlock = true
                if isFirst { result.append(blockEnclose.0); isFirst = false }
                text = text.remove(format)
                text = text.trim().enclose(lineEnclose)
            } else if isBlock {
                isBlock = false
                isFirst = true
                text = text.append(blockEnclose.1+"\n")
            }
            result.append(text)
        }

        if isBlock { result.append(blockEnclose.1) } // close open blocks
        
        md = NSMutableString(string: result.joined(separator: "\n"))
    }
*/    
}


// End

File: Public/static/markdown.js
// Markdown.js

function Markdown(text) {
	//console.log("Markdown enter...")
	var md = removeCR(text)

    md = cleanHtml(md)
    md = parseHeaders(md)
    md = parseBold(md)
    md = parseItalic(md)
    md = parseDeleted(md)
    md = parseImages(md)
    md = parseLinks(md)
    md = parseYoutubeVideos(md)
    md = parseCodeBlock(md)
    md = parseCodeInline(md)
    md = parseHorizontalRule(md)
    md = parseUnorderedLists(md)
    md = parseOrderedLists(md)
    md = parseBlockquotes(md)
    md = parseParagraphs(md)

	//console.log("Markdown exit...")
    return md

    // Internal methods
	function removeCR(text) {
		return text.split("\r\n").join("\n")
	}

	function cleanHtml(md) {
	    return md.replace(/<.*?>/gm, "")
	}

	function parseHeaders(md) {
		var mx = md
	    mx = mx.replace(/^###(.*)?$/m, "<h3>$1</h3>")
	    mx = mx.replace(/^##(.*)?$/m, "<h2>$1</h2>")
	    mx = mx.replace(/^#(.*)?$/m, "<h1>$1</h1>")
	    return mx
	}

	function parseBold(md) {
	    return md.replace(/\*\*(.*?)\*\*/gm, "<b>$1</b>")
	}

	function parseItalic(md) {
	    return md.replace(/\*(.*?)\*/gm, "<i>$1</i>")
	}

	function parseDeleted(md) {
	    return md.replace(/~~(.*?)~~/gm, "<s>$1</s>")
	}

	function parseImages(md) {
		var mx = md
	    mx = mx.replace(/!\[(\d+)x(\d+)\]\((.*?)\)/gm, "<img src=\"$3\" width=\"$1\" height=\"$2\" />")
	    mx = mx.replace(/!\[(.*?)\]\((.*?)\)/gm, "<img alt=\"$1\" src=\"$2\" />")
	    return mx
	}

	function parseLinks(md) {
		var mx = md
	    mx = mx.replace(/\[(.*?)\]\((.*?)\)/gm, "<a href=\"$2\">$1</a>")
	    mx = mx.replace(/\[http(.*?)\]/gm, "<a href=\"http$1\">http$1</a>")
	    mx = mx.replace(/(^|\s)http(.*?)(\s|\.\s|\.$|,|$)/gm, "$1<a href=\"http$2\">http$2</a>$3")
	    return mx
	}

	function parseCodeBlock(md) {
	    return md.replace(/\`\`\`\n?([^`]+)\`\`\`/gm, "<pre>$1</pre>")
	}

	function parseCodeInline(md) {
	    return md.replace(/\`(.*?)\`/gm, "<code>$1</code>")
	}

	function parseHorizontalRule(md) {
	    return md.replace(/^---/gm, "<hr>")
	}

	function parseUnorderedLists(md) {
		var rex = /^\s*\*/gm
	    return parseBlock(md, rex, "<ul>", "</ul>", "<li>", "</li>")
	}

	function parseOrderedLists(md) {
		var rex = /^\s*\d+[\.|-]/gm
	    return parseBlock(md, rex, "<ol>", "</ol>", "<li>", "</li>")
	}

	function parseBlockquotes(md) {
		var rex1 = /^>/gm
		var rex2 = /^:/gm
		var mx = md
	    mx = parseBlock(mx, rex1, "<blockquote>", "</blockquote>")
	    mx = parseBlock(mx, rex2, "<blockquote>", "</blockquote>")
	    return mx
	}

	function parseYoutubeVideos(md) {
	    return md.replace(/\[youtube (.*?)\]/gm, "<p><a href=\"http://www.youtube.com/watch?v=$1\" target=\"_blank\"><img src=\"http://img.youtube.com/vi/$1/0.jpg\" alt=\"Youtube video\" width=\"240\" height=\"180\" /></a></p>")
	}

	function parseParagraphs(md) {
	    return md.replace(/\n\n([^\n]+)\n\n/gm, "\n\n<p>$1</p>\n\n")
	}

	function parseBlock(md, pattern, blockIni, blockEnd, lineIni, lineEnd) {
		if(!lineIni){ lineIni = "" }
		if(!lineEnd){ lineEnd = "" }
	    var lines = md.split("\n")
	    var result = []
	    var isBlock = false
	    var isFirst = true
	    
	    for(i in lines) {
	        var text = lines[i]
	        if(text.match(pattern)) {
	            isBlock = true
	            if(isFirst) { result.push(blockIni); isFirst = false }
	            text = text.replace(pattern, "")
	            text = lineIni + text.trim() + lineEnd
	        } else if(isBlock) {
	            isBlock = false
	            isFirst = true
	            text += blockEnd+"\n"
	        }
	        result.push(text)
	    }

	    if(isBlock) { result.push(blockEnd) } // close open blocks
	    
	    var mx = result.join("\n")

	    return mx
	}
}

// End